source/lac/trilinos_sparse_matrix.cc

```diff
@@ -2439,6 +2443,784 @@ namespace TrilinosWrappers
+                            "(Default constructor)"));
+        };
+
+        inv_Tvmult = [](Range &, const Domain &)
+        {
+          Assert(false,
+                 ExcMessage("Uninitialized TrilinosPayload::inv_Tvmult called"
+                            "(Default constructor)"));
+        };
+      }
+
+
+
+      TrilinosPayload::TrilinosPayload (const TrilinosWrappers::SparseMatrix &matrix_exemplar,
+                                        const TrilinosWrappers::SparseMatrix &matrix)
+        : use_transpose (matrix_exemplar.trilinos_matrix().UseTranspose()),
+          communicator (matrix_exemplar.get_mpi_communicator()),
+          domain_map (matrix_exemplar.locally_owned_domain_indices().make_trilinos_map(communicator.Comm())),
+          range_map (matrix_exemplar.locally_owned_range_indices().make_trilinos_map(communicator.Comm()))
+      {
+        vmult = [&matrix_exemplar,&matrix](Range &tril_dst, const Domain &tril_src)
+        {
+          // Duplicated from TrilinosWrappers::SparseMatrix::vmult
+          Assert (&tril_src != &tril_dst, TrilinosWrappers::SparseMatrix::ExcSourceEqualsDestination());
+          Assert (matrix.trilinos_matrix().Filled(), TrilinosWrappers::SparseMatrix::ExcMatrixNotCompressed());
+          internal::check_vector_map_equality(matrix_exemplar.trilinos_matrix(),
+                                              tril_src, tril_dst,
+                                              matrix_exemplar.trilinos_matrix().UseTranspose());
+          internal::check_vector_map_equality(matrix.trilinos_matrix(),
+                                              tril_src, tril_dst,
+                                              matrix.trilinos_matrix().UseTranspose());
+
+          const int ierr = matrix.trilinos_matrix().Apply (tril_src, tril_dst);
+          AssertThrow (ierr == 0, ExcTrilinosError(ierr));
+        };
+
+        Tvmult = [&matrix_exemplar,&matrix](Domain &tril_dst, const Range &tril_src)
+        {
+          // Duplicated from TrilinosWrappers::SparseMatrix::Tvmult
+          Assert (&tril_src != &tril_dst, TrilinosWrappers::SparseMatrix::ExcSourceEqualsDestination());
+          Assert (matrix.trilinos_matrix().Filled(), TrilinosWrappers::SparseMatrix::ExcMatrixNotCompressed());
+          internal::check_vector_map_equality(matrix_exemplar.trilinos_matrix(),
+                                              tril_src, tril_dst,
+                                              !matrix_exemplar.trilinos_matrix().UseTranspose());
+          internal::check_vector_map_equality(matrix.trilinos_matrix(),
+                                              tril_src, tril_dst,
+                                              !matrix.trilinos_matrix().UseTranspose());
+
+          Epetra_CrsMatrix &tril_mtrx_non_const = const_cast<Epetra_CrsMatrix &>(matrix.trilinos_matrix());
+          tril_mtrx_non_const.SetUseTranspose(!matrix.trilinos_matrix().UseTranspose());
```

Comment:
What will happen if you have two linear operator objects pointing to the same matrix, one for just the matrix and one for the transpose? You do reset the flag on the matrix below, but I suppose this makes it impossible to use these objects at the same time on different threads...